File: Best Time to Buy and Sell Stock (109)/code.cpp
#include <iostream>
#include <vector>
#include <limits>
using namespace std;

class Solution {
public:
    bool canAchieveDivisibleProfit(vector<int>& prices, int k) {
        int min_price = numeric_limits<int>::max();
        int max_profit = 0;
        
        for (int price : prices) {
            min_price = min(min_price, price);
            max_profit = max(max_profit, price - min_price);
        }
        
        return (max_profit > 0) ? (max_profit % k == 0) : false;
    }
};

int main() {
    int t;
    cin >> t;
    
    for (int i = 0; i < t; ++i) {
        int n;
        cin >> n;
        
        vector<int> prices(n);
        for (int j = 0; j < n; ++j) {
            cin >> prices[j];
        }
        
        int k;
        cin >> k;
        
        Solution solution;
        cout << (solution.canAchieveDivisibleProfit(prices, k) ? "True" : "False") << endl;
    }
    
    return 0;
}

File: Best Time to Buy and Sell Stock (109)/code.py
class Solution:
    def canAchieveDivisibleProfit(self, prices, k):
        min_price = float('inf')
        max_profit = 0
        for price in prices:
            min_price = min(min_price, price)
            max_profit = max(max_profit, price - min_price)
        return max_profit % k == 0 if max_profit > 0 else False

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        n = int(input())
        prices = list(map(int, input().split()))
        k = int(input())
        solution = Solution()
        print(solution.canAchieveDivisibleProfit(prices, k))
